### PR TITLE
Restartable engine implementation

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #85 

## Description
This PR implements restartable engine wrapper which currently supports Fibonacci sequence for the backoff time.

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

File: models/engine.go [hunk 1]
```diff
@@ -38,8 +64,30 @@ func (r *RestartableEngine) Ready() <-chan struct{} {
 }
 
 func (r *RestartableEngine) Run(ctx context.Context) error {
-	// todo add restart logic
-	return r.engine.Run(ctx)
+	var err error
+	for i, b := range r.backoff {
+		<-time.After(b) // wait for it
```
Comment (sideninja) [inline models/engine.go:69]: image

File: models/engine.go [hunk 2]
```diff
@@ -16,13 +21,34 @@ type Engine interface {
 
 var _ Engine = &RestartableEngine{}
 
+func NewRestartableEngine(engine Engine, retries int, logger zerolog.Logger) *RestartableEngine {
```
Comment (peterargue) [inline models/engine.go:24]: can you add a comment noting that if retires is less than 3, what the behavior will be

File: models/engine.go (hunk 1 above)
Comment (peterargue) [inline models/engine.go:69]: maybe add a select to exit early if the context is cancelled

**Review by peterargue — APPROVED** [review]
looks good. added a couple small comments. please also add a test covering `RestartableEngine`

File: models/engine.go (hunk 1 above)
Comment (sideninja) [reply models/engine.go:69]: oh nice yeah

File: models/engine.go (hunk 2 above)
Comment (sideninja) [reply models/engine.go:24]: lol

File: models/errors.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+func (r RecoverableError) Unwrap() error {
+	return r.err
+}
+
+func (r RecoverableError) Error() string {
+	return fmt.Sprintf("recoverable error: %v", r.err)
+}
+
+func IsRecoverableError(err error) bool {
+	return errors.As(err, &RecoverableError{})
```
Comment (m-Peter) [inline models/errors.go:31]:
```suggestion
	return errors.Is(err, RecoverableError{})
```

**Review by m-Peter — APPROVED** [review]
LGTM!

File: models/errors.go (hunk 3 above)
Comment (sideninja) [reply models/errors.go:31]: This won't work since we defined a custom error type.

Comment (sideninja) [reply models/errors.go:31]: `errors.Is` would only work if we want to compare error to a specific error value not to a type.

Comment (m-Peter) [reply models/errors.go:31]: Ah, yes, my bad :sweat_smile:

Comment (sideninja) [reply models/errors.go:31]: So if we would have a recoverable error as:
```
ErrRecoverable = errors.New("receoverable")
```
we could check with `errors.Is` but then that wouldn't work with any error, you could wrap it like: 
```
fmt.Errorf("disconnected: %w", ErrRecoverable) 
```
and that would work, but I find defining a custom type for this case a nicer pattern. But then that makes the `errors.Is` not working.

Comment (m-Peter) [reply models/errors.go:31]: Yeah, good point. I was totally off on this :see_no_evil:

Comment (sideninja) [reply models/errors.go:31]: yeah no worries, this is confusing API

**Pull request merged**